include/mumble/Base64.hpp

```diff
@@ -3,30 +3,147 @@
+		if (in.empty()) {
+			return 0;
+		}
+
+		if (in.size() % 4) {
+			throw std::runtime_error("Input is not a multiple of 4!");
+		}
+
+		const auto size = decodedSize(in);
+		if (out.size() < size) {
+			throw std::runtime_error("Insufficient output buffer size!");
+		}
+
+		for (auto data = reinterpret_cast< uint8_t * >(out.data()); !in.empty(); data += 3, in = in.subspan(4)) {
+			data[0] = static_cast< uint8_t >((charToPos(in[0]) << 2) + ((charToPos(in[1]) & 0x30) >> 4));
+
+			if (in[2] == '=') {
+				break;
+			}
+
+			data[1] = static_cast< uint8_t >(((charToPos(in[1]) & 0x0f) << 4) + ((charToPos(in[2]) & 0x3c) >> 2));
+
+			if (in[3] == '=') {
+				break;
+			}
+
+			data[2] = static_cast< uint8_t >(((charToPos(in[2]) & 0x03) << 6) + charToPos(in[3]));
+		}
+
+		return size;
+	}
+
+	static constexpr size_t encode(StrView out, const BufViewConst in) {
+		if (in.empty()) {
+			return 0;
+		}
 
-	Base64();
-	virtual ~Base64();
+		const auto size = encodedSize(in);
+		if (out.size() < size) {
+			throw std::runtime_error("Insufficient output buffer size!");
+		}
 
-	virtual explicit operator bool();
+		ByteViewConst data(reinterpret_cast< ByteViewConst::element_type * >(in.data()), in.size());
 
-	virtual size_t decode(const BufView out, const BufViewConst in);
-	static size_t encode(const BufView out, const BufViewConst in);
+		for (; !data.empty(); out = out.subspan(4)) {
+			out[0] = posToChar((data[0] & 0xfc) >> 2);
+
+			if (data.size() >= 2) {
+				out[1] =
+					posToChar(static_cast< ByteView::element_type >(((data[0] & 0x03) << 4) + ((data[1] & 0xf0) >> 4)));
+
+				if (data.size() >= 3) {
+					out[2] = posToChar(
+						static_cast< ByteView::element_type >(((data[1] & 0x0f) << 2) + ((data[2] & 0xc0) >> 6)));
+					out[3] = posToChar(data[2] & 0x3f);
+				} else {
+					out[2] = posToChar(static_cast< ByteView::element_type >((data[1] & 0x0f) << 2));
+					out[3] = PAD_CHAR;
+				}
+			} else {
+				out[1] = posToChar(static_cast< ByteView::element_type >((data[0] & 0x03) << 4));
+				out[2] = PAD_CHAR;
+				out[3] = PAD_CHAR;
+			}
+
+			// All data blocks are guaranteed to be 3 bytes except the last one.
+			data = data.subspan(std::min(size_t{ 3 }, data.size()));
+		}
+
+		return size;
+	}
+
+	static constexpr size_t decodedSize(const StrViewConst in) {
+		size_t size = (in.size() / 4) * 3;
+		// Base64 can have up to 2 pad characters, in order to make the string a multiple of 4.
+		// In order to calculate the exact output size we have to ignore padding.
+		if (auto iter = in.rbegin(); *iter == PAD_CHAR) {
+			size -= *++iter == PAD_CHAR ? 2U : 1U;
+		}
+
+		return size;
+	}
+
+	static constexpr size_t encodedSize(const BufViewConst in) { return ((in.size() + 2) / 3) * 4; }
 
 private:
-	std::unique_ptr< P > m_p;
+	using ByteView      = gsl::span< uint8_t >;
+	using ByteViewConst = gsl::span< const uint8_t >;
+
+	static constexpr uint8_t charToPos(const char ch) {
```

Comment:
```suggestion
	static constexpr uint8_t charToNumeric(const char ch) {
```
Is this a better name? `pos` seems to indicate that we were using the result to index something, which is not the case :thinking:

Reply:
Right, it's not a position in a table.

Reply:
Is `charToByte()` alright?

Reply:
That would be a perfect fit, but I think then the function should return a `std::byte` instead of a `uint8_t` :thinking: 
Afaict this shouldn't cause issues on the calling site :eyes:

Reply:
Yeah, I did that.